python/util.py

```diff
@@ -0,0 +1,257 @@
+        if obj is None:
+            return self
+        else:
+            return getattr(obj, self.identifier)
+
+    @staticmethod
+    def _limit(value, min_, max_):
+        if min_ is None and max_ is None:
+            return value
+        if min_ is None:
+            min_ = min(value, max_)
+        if max_ is None:
+            max_ = max(value, min_)
+        return sorted((value, min_, max_))[1]
+
+    def __set__(self, obj, value):
+        min_ = self.min(obj)
+        max_ = self.max(obj)
+        setattr(obj, self.identifier, self._limit(value, min_, max_))
+
+UNASSIGNED = object()
+
+
+class AB:
+    '''Vector like class to help process multiple related values together.
+    Provides useful functionality like mapped attribute lookups and
+    distributive calling.
+    '''
+
+    __slots__ = '_data',
+
+    def __init__(self, a, b=UNASSIGNED):
+        'Takes either values for a and b or a value to repeat for both'
+        if b is UNASSIGNED:
+            b = a
+        self._data = [a, b]
+
+    def __repr__(self):
+        return '{}({}, {})'.format(self.__class__.__name__, *self)
+
+    def __iter__(self):
+        return iter(self._data)
+
+    def __len__(self):
+        return len(self._data)
+
+    def __getitem__(self, i):
+        return self._data[i]
+
+    def __setitem__(self, i, value):
+        self._data[i] = value
+
+    def __getattr__(self, name):
+        return self.map(lambda obj: getattr(obj, name))
+
+    def __call__(self, *args, **kwargs):
+        return self.map(lambda obj: obj(*args, **kwargs))
+
+    def distribute(self, *iterables):
+        '''Like ab(arg1, arg2...), but we distribute the (iterable) arguments to
+        each respective ab:
+          ab[0](arg1[0], arg2[0]...)
+          ab[1](arg1[1], arg2[1]...)
+        '''
+        iterators = tuple(map(iter, iterables))
+        f_on_next_i = lambda f: f(*map(next, iterators))
+        return self.map(f_on_next_i)
+
+    def __add__(self, other):
+        return self.__class__(*starmap(add, zip(self, other)))
+
+    def __sub__(self, other):
+        return self.__class__(*starmap(sub, zip(self, other)))
+
+    def __eq__(self, other):
+        return len(self) == len(other) and all(starmap(eq, zip(self, other)))
+
+    a = property(lambda self: self[0])
+    b = property(lambda self: self[1])
+
+    @classmethod
+    def from_map(cls, func, iterable):
+        return cls(*map(func, iterable))
+
+    def map(self, func, *args, **kwargs):
+        return self.from_map(lambda i: func(i, *args, **kwargs), self)
+
+    def reversed(self):
+        return self.__class__(*reversed(self))
+
+
+class FormattedChar:
+    __slots__ = 'pre_fmt', 'c', 'post_fmt'
+
+    def __init__(self, c, pre_fmt='', post_fmt=''):
+        self.pre_fmt = pre_fmt
+        self.c = c
+        self.post_fmt = post_fmt
+
+    def __str__(self):
+        return self.pre_fmt + self.c + self.post_fmt
+
+    def __repr__(self):
+        return 'c({!r}, {!r}, {!r})'.format(
```

Comment:
I was using the repr mainly for debugging and it looked cluttered. I might be replacing this way of handling formatting anyway - watch this space!